### PR TITLE
Remove `GLWrapper` and move implementation to `IRenderer`

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu-framework/pull/5330

For the most part, this is a 1-1 migration of `GLWrapper` implementation into `OpenGLRenderer`, with appropriate interfaces added to `IRenderer`. The rest is passing `IRenderer`/`OpenGLRenderer` to GL-specific classes (VBOs, Shaders, etc).

Important changes of note:
- `Drawable` DIs in an `IRenderer` to dispose `DrawNode`s. Somehow, `DrawNode` disposal has to be enqueued to run 3 frames behind. There's several ways to do this, and I'm not sure if the way I've chosen here is the best. Alternatives:
    1. Make the disposal queue static.
        * Reason not done: sounds unsafe.
    2. Change `Drawable.CreateDrawNode()` to take an `IRenderer` parameter and construct `DrawNode`s with the renderer.
        * Reason not done: larger breaking change.
    3. Store `IRenderer` from `Drawable.Draw()`/`Drawable.DrawOpaqueInteriorSubTree()`.
        * Reason not done: I'm not sure if this will work if exceptions are thrown (e.g. `BufferedDrawNode()` calls `base.Draw()` very late.
- `TexturedShaderDrawNode` interface has changed slightly because `RequiresRoundedShader()` needs an `IRenderer` parameter. osu!-side this only results in a changes to one file.
- Size check removed from `TextureUpload`. In #5330 I've moved the check to `Texture.SetData()`.

# vNext

## `GLWrapper` removed, replaced by `IRenderer`

The static `GLWrapper` class no longer exists, and all drawing functions should be performed via the `DrawNode.Draw()` `IRenderer` parameter. There is a 1-1 compatibility between the classes.

## `TexturedShaderDrawNode` interface changed slightly

- `Shader` -> `GetAppropriateShader(IRenderer)`.
- `RequriesRoundedShader` -> `RequiresRoundedShader(IRenderer)`

### CODE REVIEW

**Comment by smoogipoo** [comment]
Will need a test on `iOS`, but I'm fairly confident it should work.

File: osu.Framework/Graphics/Rendering/IRenderer.cs [hunk 1]
```diff
@@ -15,12 +18,70 @@ namespace osu.Framework.Graphics.Rendering
...
+
+        /// <summary>
+        /// The maximum number of texture uploads to dequeue and upload per frame.
+        /// Defaults to 32.
+        /// </summary>
+        int MaxTexturesUploadedPerFrame { get; set; }
+
+        /// <summary>
+        /// The maximum number of pixels to upload per frame.
+        /// Defaults to 2 megapixels (8mb alloc).
```
Comment (peppy) [inline osu.Framework/Graphics/Rendering/IRenderer.cs:46]: Kinda weird to specify the defaults in the interface when it doesn't guarantee them, but I guess it's okay.

**Comment by peppy** [comment]
While testing on iOS, I got some crashes:

JetBrains Rider 2022-08-05 at 07 04 54

and

```csharp
System.NullReferenceException: Object reference not set to an instance of an object
  at osu.Framework.Graphics.OpenGL.Buffers.FrameBuffer+FrameBufferTexture.set_Width (System.Int32 value) [0x00000] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/OpenGL/Buffers/FrameBuffer.cs:136 
  at osu.Framework.Graphics.OpenGL.Textures.TextureGL..ctor (osu.Framework.Graphics.OpenGL.OpenGLRenderer renderer, System.Int32 width, System.Int32 height, System.Boolean manualMipmaps, osuTK.Graphics.ES30.All filteringMode, SixLabors.ImageSharp.PixelFormats.Rgba32 initialisationColour) [0x00025] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/OpenGL/Textures/TextureGL.cs:66 
  at osu.Framework.Graphics.OpenGL.Buffers.FrameBuffer+FrameBufferTexture..ctor (osu.Framework.Graphics.OpenGL.OpenGLRenderer renderer, osuTK.Graphics.ES30.All filteringMode) [0x00000] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/OpenGL/Buffers/FrameBuffer.cs:125 
  at osu.Framework.Graphics.OpenGL.Buffers.FrameBuffer..ctor (osu.Framework.Graphics.OpenGL.OpenGLRenderer renderer, osuTK.Graphics.ES30.RenderbufferInternalFormat[] renderBufferFormats, osuTK.Graphics.ES30.All filteringMode) [0x00030] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/OpenGL/Buffers/FrameBuffer.cs:27 
  at osu.Framework.Graphics.OpenGL.OpenGLRenderer.CreateFrameBuffer (osu.Framework.Graphics.Rendering.RenderBufferFormat[] renderBufferFormats, osu.Framework.Graphics.Textures.TextureFilteringMode filteringMode) [0x000a5] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/OpenGL/OpenGLRenderer.cs:855 
  at osu.Framework.Graphics.BufferedDrawNodeSharedData.Initialise (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x00021] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/BufferedDrawNodeSharedData.cs:92 
  at osu.Framework.Graphics.BufferedDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x00013] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/BufferedDrawNode.cs:87 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw (osu.Framework.Graphics.Rendering.IRenderer renderer) [0x0009d] in /Users/dean/Projects/osu-framework/osu.Framework/Graphics/Containers/CompositeDrawable_DrawNode.cs:204 
  at osu.Framework.Platform.GameHost.DrawFrame () [0x00168] in /Users/dean/Projects/osu-framework/osu.Framework/Platform/GameHost.cs:506 
  at osu.Framework.Threading.GameThread.processFrame () [0x000d9] in /Users/dean/Projects/osu-framework/osu.Framework/Threading/GameThread.cs:454 
--- End of stack trace from previous location where exception was thrown ---

  at osu.Framework.Platform.GameHost+<>c__DisplayClass133_0.<abortExecutionFromException>b__0 () [0x00002] in /Users/dean/Projects/osu-framework/osu.Framework/Platform/GameHost.cs:390 
  at osu.Framework.Threading.ScheduledDelegate.InvokeTask () [0x00010] in /Users/dean/Projects/osu-framework/osu.Framework/Threading/ScheduledDelegate.cs:106 
  at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal () [0x0002b] in /Users/dean/Projects/osu-framework/osu.Framework/Threading/ScheduledDelegate.cs:92 
  at osu.Framework.Threading.Scheduler.Update () [0x0008a] in /Users/dean/Projects/osu-framework/osu.Framework/Threading/Scheduler.cs:125 
  at osu.Framework.Threading.GameThread.processFrame () [0x00172] in /Users/dean/Projects/osu-framework/osu.Framework/Threading/GameThread.cs:467 
  at osu.Framework.Threading.GameThread.RunSingleFrame () [0x00001] in /Users/dean/Projects/osu-framework/osu.Framework/Threading/GameThread.cs:296 
  at osu.Framework.Platform.ThreadRunner.RunMainLoop () [0x0005f] in /Users/dean/Projects/osu-framework/osu.Framework/Platform/ThreadRunner.cs:114 
  at osu.Framework.Platform.GameHost.windowUpdate () [0x00026] in /Users/dean/Projects/osu-framework/osu.Framework/Platform/GameHost.cs:850 
  at osu.Framework.Platform.GameHost.<Run>b__155_2 (System.Object _, osuTK.FrameEventArgs _) [0x00000] in /Users/dean/Projects/osu-framework/osu.Framework/Platform/GameHost.cs:754 
  at (wrapper delegate-invoke) System.EventHandler`1[osuTK.FrameEventArgs].invoke_void_object_TEventArgs(object,osuTK.FrameEventArgs)
  at osuTK.iOS.iOSGameView.OnUpdateFrame (osuTK.FrameEventArgs e) [0x0000a] in <0fc4a59399e2486f977f9fdb230650a4>:0 
  at osuTK.iOS.iOSGameView.RunIteration (Foundation.NSTimer timer) [0x00042] in <0fc4a59399e2486f977f9fdb230650a4>:0 
  at osuTK.iOS.CADisplayLinkTimeSource.RunIteration (Foundation.NSObject parameter) [0x00000] in <0fc4a59399e2486f977f9fdb230650a4>:0 
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.Type principalClass, System.Type delegateClass) [0x0003b] in /Users/builder/azdo/_work/1/s/xamarin-macios/src/UIKit/UIApplication.cs:83 
  at osu.Framework.Tests.Application.Main (System.String[] args) [0x00001] in /Users/dean/Projects/osu-framework/osu.Framework.Tests.iOS/Application.cs:18 
```

JetBrains Rider 2022-08-05 at 07 07 45

JetBrains Rider 2022-08-05 at 07 07 57

Both of these only started after selecting `osu.Framework.Tests.Visual.Drawables.TestSceneCircularArcBoundingBox`, and now occur on startup due to that test being loaded first.

**Comment by smoogipoo** [comment]
Good catch, came about due to the removal of the null check in the previous PR. Fixed.

**Comment by peppy** [comment]
I guess the first crash was just triggered as a side-effect of the second and in normal execution won't happen due to the `Renderer` being the first thing to be initialised now?

**Comment by smoogipoo** [comment]
That's correct, yeah.

**Review by peppy — APPROVED** [review]

**Pull request merged**